service/udp.go

```diff
@@ -321,92 +320,239 @@ func (c *natconn) ReadFrom(buf []byte) (int, net.Addr, error) {
 // Packet NAT table
 type natmap struct {
 	sync.RWMutex
-	keyConn map[string]*natconn
-	logger  *slog.Logger
-	timeout time.Duration
-	metrics UDPMetrics
+	associations map[string]PacketAssociation
 }
 
-func newNATmap(timeout time.Duration, sm UDPMetrics, l *slog.Logger) *natmap {
-	m := &natmap{logger: l, metrics: sm}
-	m.keyConn = make(map[string]*natconn)
-	m.timeout = timeout
-	return m
+func newNATmap() *natmap {
+	return &natmap{associations: make(map[string]PacketAssociation)}
 }
 
-func (m *natmap) Get(key string) *natconn {
+// Get returns a UDP NAT entry from the natmap.
+func (m *natmap) Get(clientAddr string) PacketAssociation {
 	m.RLock()
 	defer m.RUnlock()
-	return m.keyConn[key]
+	return m.associations[clientAddr]
 }
 
-func (m *natmap) set(key string, pc net.PacketConn, cryptoKey *shadowsocks.EncryptionKey, connMetrics UDPConnMetrics) *natconn {
-	entry := &natconn{
-		PacketConn:     pc,
-		cryptoKey:      cryptoKey,
-		metrics:        connMetrics,
-		defaultTimeout: m.timeout,
+// Del deletes a UDP NAT entry from the natmap.
+func (m *natmap) Del(clientAddr string) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, ok := m.associations[clientAddr]; ok {
+		delete(m.associations, clientAddr)
 	}
+}
 
+// Add adds a new UDP NAT entry to the natmap.
+func (m *natmap) Add(clientAddr string, assoc PacketAssociation) {
 	m.Lock()
 	defer m.Unlock()
 
-	m.keyConn[key] = entry
-	return entry
+	m.associations[clientAddr] = assoc
 }
 
-func (m *natmap) del(key string) net.PacketConn {
+func (m *natmap) Close() error {
 	m.Lock()
 	defer m.Unlock()
 
-	entry, ok := m.keyConn[key]
-	if ok {
-		delete(m.keyConn, key)
-		return entry
+	var err error
+	for _, assoc := range m.associations {
+		if e := assoc.Close(); e != nil {
+			err = e
+		}
 	}
-	return nil
+	return err
 }
 
-func (m *natmap) Add(clientAddr net.Addr, clientConn net.PacketConn, cryptoKey *shadowsocks.EncryptionKey, targetConn net.PacketConn, keyID string) *natconn {
-	connMetrics := m.metrics.AddUDPNatEntry(clientAddr, keyID)
-	entry := m.set(clientAddr.String(), targetConn, cryptoKey, connMetrics)
+// ConnAssociation represents a UDP association that handles incoming and
+// outgoing packets from a [net.Conn] and forwards them to a target connection,
+// and vice versa. Used by Caddy.
+type ConnAssociation interface {
```

Comment:
Delete this interface, it's not needed and is confusing in the presence of PacketAssociation. PacketAssociation is the only one you need, since the ConnAssociation.Handle can be implemented from PacketAssociation.